### PR TITLE
fix(awscdk): issues with CDK v2 libraries and apps (take 2)

### LINKED ISSUE #1173: 1-buster-slim breaks workflows using cdk v2.0.0-rc.25
My PR's build is failing with
```
error @aws-cdk/assert@2.0.0-rc.25: The engine "node" is incompatible with this module. Expected version ">= 14.15.0". Got "12.22.7"
```
while trying to upgrade All The Things in my corp `cdk-library`. 
```
❯ npx projen --version
0.27.51
```
One of the things I see is that the upgrade swapped:
```js
      image: jsii/superchain:node14
      image: jsii/superchain:1-buster-slim
```
in the workflows. I don't know what `1-buster-slim` is, but... it looks like it's breaking stuff.

Workaround in `.projenrc.js`:
```js
  workflowContainerImage: 'jsii/superchain:node14', // Workaround buster-slim issue
```

### LINKED ISSUE #1288: AWS CDK v2 Support
CDK v2 has been released!

https://aws.amazon.com/about-aws/whats-new/2021/12/aws-cloud-development-kit-cdk-generally-available/

Would it be possible to initialize a new CDK v2 project using projen?

### LINKED ISSUE #1326: AwsCdkTypeScriptApp still uses deprecated assert library
The AwsCdkTypeScriptApp still has a [hard coded](https://github.com/projen/projen/blob/main/src/awscdk/awscdk-app-ts.ts#L106) do the deprecated assert library. 

To keep it kind of compatible, I'd move that line to the CDK v1 block, and add the assertion library in the v2 block. 

I can provide a pull request, but would like to discuss how it should look like as it might be a breaking change.

### PR DESCRIPTION
**Pull request by eladb**
Extract the logic that handles CDK dependencies (and supports v1/v2 differences) into a new component called `AwsCdkDeps` and reuse across  `AwsCdkTypeScriptApp` and `AwsCdkConstructLibrary`.

This is a follow up on #1327 which we had to revert due to breaking upgrade flows. The difference is that this change does NOT introduce breaking behavioral change to `cdkVersion` and dev/peer dependencies of constructs.

* Modify `LambdaFunction` to determine how to render import statements in the generated function construct source code.
* Fix the sample test code in `awscdk-app-ts` to work with the new assertions library.
* Change the default CDK version for `awscdk-app-ts` to `2.1.0`.
* Determines workflow container image based on CDK version

Fixes #1288
Fixes #1326
Fixes #1173

BREAKING CHANGE: `LambdaFunction` and `AutoDiscover` now requires `cdkDeps` so they can interact with CDK version manager.

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/awscdk/awscdk-deps.ts [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+  public readonly cdkMinimumVersion: string;
+
+  /**
+   * Whether CDK dependencies are added as normal dependencies (and peer dependencies).
+   * @deprecated Not used for CDK 2.x
+   */
+  public readonly cdkDependenciesAsDeps: boolean;
+
+  /**
+   * The major version of the AWX CDK (e.g. 1, 2, ...)
```
Comment (Chriscbr) [inline src/awscdk/awscdk-deps.ts:112]:
```suggestion
   * The major version of the AWS CDK (e.g. 1, 2, ...)
```

File: src/awscdk/awscdk-deps.ts [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+        break;
+    }
+
+    this.project.deps.addDependency(`constructs@${versionRequirement}`, this.dependencyType);
+
+    return versionRequirement;
+  }
+
+  /**
+   * Addsa depdenencies on the AWS CDK framework (e.g. `@aws-cdk/core` for V1 or `aws-cdk-lib` for V1).
```
Comment (Chriscbr) [inline src/awscdk/awscdk-deps.ts:217]:
```suggestion
   * Adds a dependency on the AWS CDK framework (e.g. `@aws-cdk/core` for V1 or `aws-cdk-lib` for V1).
```

File: src/awscdk/awscdk-deps.ts [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+    }
+
+    const testDeps = new Array<string>();
+
+    if (options.cdkAssert ?? true) {
+      testDeps.push('@aws-cdk/assert');
+    }
+
+    // @aws-cdk/assertions is only available starting v1.111.0
+    if (semver.gte(this.cdkMinimumVersion, '1.111.0') && (options.cdkAssertions ?? true)) {
```
Comment (Chriscbr) [inline src/awscdk/awscdk-deps.ts:264]: Good check 👍

File: src/javascript/node-package.ts [hunk 4]
```diff
@@ -838,6 +838,12 @@ export class NodePackage extends Component {
     if (pinned) {
       for (const dep of this.project.deps.all.filter(d => d.type === DependencyType.PEER)) {
         let req = dep.name;
+
+        // skip if we already have a runtime dependency on this peer
+        if (this.project.deps.tryGetDependency(dep.name, DependencyType.RUNTIME)) {
+          continue;
+        }
```
Comment (Chriscbr) [inline src/javascript/node-package.ts:845]: Feels like this should be a unit test

**Review by Chriscbr — APPROVED** [review]

File: src/javascript/node-package.ts (hunk 4 above)
Comment (eladb) [reply src/javascript/node-package.ts:845]: I added one

**Pull request merged**